packages/block-library/src/query-title/edit.js

```diff
@@ -12,43 +12,91 @@ import {
 	BlockControls,
 	useBlockProps,
 	Warning,
+	RichText,
 } from '@wordpress/block-editor';
-import { __ } from '@wordpress/i18n';
+import { __, _x } from '@wordpress/i18n';
+import { useEffect } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import HeadingLevelDropdown from '../heading/heading-level-dropdown';
 
-const SUPPORTED_TYPES = [ 'archive' ];
+const SUPPORTED_TEMPLATES = [ 'archive', 'search', '404' ];
 
 export default function QueryTitleEdit( {
-	attributes: { type, level, textAlign },
+	attributes: { content, level, textAlign },
 	setAttributes,
+	context: { templateSlug },
 } ) {
 	const TagName = `h${ level }`;
 	const blockProps = useBlockProps( {
-		className: classnames( {
+		className: classnames( 'wp-block-query-title__placeholder', {
 			[ `has-text-align-${ textAlign }` ]: textAlign,
-			'wp-block-query-title__placeholder': type === 'archive',
 		} ),
 	} );
-	// The plan is to augment this block with more
-	// block variations like `Search Title`.
-	if ( ! SUPPORTED_TYPES.includes( type ) ) {
+	let titleElement;
+
+	// Infer title content from template slug context
+	// Defaults to content attribute prop
+	switch ( templateSlug ) {
+		case 'archive':
+			// translators: Title for archive template.
+			content = _x( 'Archive title', 'archive template title' );
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
+			break;
+		case 'search':
+			// translators: Title for search template with dynamic content placeholders.
+			const placeholderContent = _x(
+				'Search results for "%search%"',
+				'search template title'
+			);
+
+			// Use placeholder content if new Query Title block is added
+			if ( content === '' ) {
+				content = placeholderContent;
+			}
+
+			titleElement = (
+				<div { ...blockProps }>
+					<RichText
+						tagName={ TagName }
+						value={ content }
+						placeholder={ placeholderContent }
+						allowedFormats={ [ 'core/bold', 'core/italic' ] }
+						onChange={ ( newContent ) =>
+							setAttributes( { content: newContent } )
+						}
+						disableLineBreaks={ true }
+					/>
+				</div>
+			);
+			break;
+		case '404':
+			// translators: Title for 404 template.
+			content = _x( 'Nothing found', '404 template title' );
+			titleElement = <TagName { ...blockProps }>{ content }</TagName>;
```

Comment:
IMO this should be editable. I don't think every user should want `Nothing found` as a title without the ability to change it.

Reply:
If they want to edit it can't they just add their own heading block?